x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java

```diff
@@ -766,41 +767,105 @@ private static void checkRemoteEnrich(LogicalPlan plan, Set<Failure> failures) {
     }
 
     /**
-     * Checks whether a condition contains a disjunction with the specified typeToken. Adds to failure if it does.
+     * Checks whether a condition contains a disjunction with a full text search.
+     * If it does, check that every element of the disjunction is a full text search or combinations (AND, OR, NOT) of them.
+     * If not, add a failure to the failures collection.
      *
-     * @param condition        condition to check for disjunctions
+     * @param condition        condition to check for disjunctions of full text searches
      * @param typeNameProvider provider for the type name to add in the failure message
      * @param failures         failures collection to add to
      */
-    private static void checkNotPresentInDisjunctions(
+    private static void checkFullTextSearchDisjunctions(
         Expression condition,
         java.util.function.Function<FullTextFunction, String> typeNameProvider,
         Set<Failure> failures
     ) {
-        condition.forEachUp(Or.class, or -> {
-            checkNotPresentInDisjunctions(or.left(), or, typeNameProvider, failures);
-            checkNotPresentInDisjunctions(or.right(), or, typeNameProvider, failures);
+        int failuresCount = failures.size();
+        condition.forEachDown(Or.class, or -> {
+            if (failures.size() > failuresCount) {
+                // Exit early if we already have a failures
+                return;
+            }
+            Expression left = or.left();
```

Comment:
Is my understanding right, that the goal of `checkFullTextSearchDisjunctions` is - if there is `FullTextFunction` exists under `Or`, all of the children of the `Or` has to be `FullTextFunction`s? If this is true, can this algorithm be simplified, instead of checking `onlyFullTextFunctionsInExpression`, we can check `noneFullTextFunctionExists` on both sides of the `Or`?

Reply:
Sorry, I'm not following - can you please elaborate with an example?

Reply:
I'm thinking if we can simplify the logic here. Will something like this work? Do we need to check both sides separately?

```
            boolean hasFullText = or.anyMatch(FullTextFunction.class::isInstance);
            boolean hasOnlyFullText = onlyFullTextFunctionsInExpression(or);
            if (hasFullText) {
                if (hasOnlyFullText) {
                    // succeed
                } else {
                    // fail
                }
            } else {
                // succeed
            }
```